### PR TITLE
[WX-963] Unzip Engine Function 

### PR DESCRIPTION
**Pull request by THWiseman**
Implemented the [unzip()](https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md#-pairarrayx-arrayy-unziparraypairx-y) engine function, new with WDL 1.1. 

As a new WDL 1.1 feature, it was implemented for `Biscayne` and `Cascades`, but not earlier versions (e.g. `draft-2`, `draft-3`).

### CODE REVIEW

**Review by aednichols — APPROVED** [review]
Nice!

File: wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/values/CascadesValueEvaluatorSpec.scala [hunk 1]
```diff
@@ -217,4 +217,30 @@ class CascadesValueEvaluatorSpec extends AnyFlatSpec with CromwellTimeoutSpec wi
...
+    val right: WomArray = WomArray(WomArrayType(WomIntegerType), Seq(WomInteger(1), WomInteger(2), WomInteger(3)))
+    val expectedPair: WomPair = WomPair(left, right)
+
+    expr.shouldBeValidPF { case e =>
+      e.evaluateValue(Map.empty, NoIoFunctionSet, None) shouldBeValid EvaluatedValue(expectedPair, Seq.empty)
+    }
+  }
+
+  it should "evaluate an unzip on an empty collection correctly" in {
+    val str = """ unzip([])"""
```
Comment (jgainerdewar) [inline wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/values/CascadesValueEvaluatorSpec.scala:235]: Is it worth testing `unzip([()])`?

File: wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/types/CascadesTypeEvaluatorSpec.scala [hunk 2]
```diff
@@ -64,4 +64,12 @@ class CascadesTypeEvaluatorSpec extends AnyFlatSpec with CromwellTimeoutSpec wit
       e.evaluateType(Map.empty) shouldBeValid WomArrayType(WomStringType)
     }
   }
+
```
Comment (jgainerdewar) [inline wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/types/CascadesTypeEvaluatorSpec.scala:67]: There's another test in the Biscayne version, should it be added here?

**Review by jgainerdewar — COMMENTED** [review]
Looks great! 

Reviewing this made me think about how we don't seem to have much unit testing of failure modes (my suffix PR suffered from the same issue) - what if I run a WDL with `unzip([("banana", 5), (6, "apple"])`? Is that valid, or does it fail because we can't find good types for the output arrays?

**Comment by THWiseman** [comment]
> Looks great!
> 
> Reviewing this made me think about how we don't seem to have much unit testing of failure modes (my suffix PR suffered from the same issue) - what if I run a WDL with `unzip([("banana", 5), (6, "apple")])`? Is that valid, or does it fail because we can't find good types for the output arrays?

Agree with this sentiment! I added a few unit tests for basic failure modes.

To your more specific question: `unzip([("banana", 5), (6, "apple")])` will *succeed*, because putting stuff into arrays homogenizes the type, and the integers are coercible to strings. For that to fail, I think we would want arrays to throw errors instead of attempting to homogenize types, which is a big and breaking change that we probably shouldn't make. 

However, unzip `([(1, 2), ([1,2,3], 3)])` will fail because there's no way to coerce between `int` and `array[int]`. Added a test case to assert this.

File: wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/types/CascadesTypeEvaluatorSpec.scala (hunk 2 above)
Comment (jgainerdewar) [reply wdl/transforms/cascades/src/test/scala/wdl/transforms/cascades/linking/expression/types/CascadesTypeEvaluatorSpec.scala:67]: I think we're still missing a test here that is is Biscayne, no? Once that's in I think this is good to go!

**Review by jgainerdewar — APPROVED** [review]

**Pull request merged**